src/bn254/ec_add/mod.rs

```diff
@@ -0,0 +1,330 @@
+use boojum::gadgets::queue::CircuitQueueWitness;
+use boojum::gadgets::queue::QueueState;
+use boojum::gadgets::traits::allocatable::{CSAllocatableExt, CSPlaceholder};
+use boojum::gadgets::traits::round_function::CircuitRoundFunction;
+use boojum::gadgets::traits::selectable::Selectable;
+use boojum::gadgets::traits::witnessable::WitnessHookable;
+use boojum::gadgets::u160::UInt160;
+use boojum::gadgets::u256::UInt256;
+use boojum::gadgets::u32::UInt32;
+use boojum::gadgets::u8::UInt8;
+use boojum::pairing::CurveAffine;
+use cs_derive::*;
+use derivative::Derivative;
+use zkevm_opcode_defs::system_params::PRECOMPILE_AUX_BYTE;
+
+use crate::base_structures::log_query::*;
+use crate::base_structures::memory_query::*;
+use crate::base_structures::precompile_input_outputs::PrecompileFunctionOutputData;
+use crate::bn254::ec_add::input::EcAddCircuitInputOutput;
+use crate::demux_log_queue::StorageLogQueue;
+use crate::ethereum_types::U256;
+use crate::fsm_input_output::circuit_inputs::INPUT_OUTPUT_COMMITMENT_LENGTH;
+use crate::fsm_input_output::*;
+use crate::sha256_round_function::input::Sha256RoundFunctionFSM;
+use crate::storage_application::ConditionalWitnessAllocator;
+
+use super::*;
+
+use self::ec_mul::implementation::{
+    convert_field_element_to_uint256, convert_uint256_to_field_element,
+};
+use self::input::EcAddCircuitInstanceWitness;
+
+pub mod implementation;
+pub mod input;
+
+pub const MEMORY_QUERIES_PER_CALL: usize = 4;
+pub const NUM_MEMORY_READS_PER_CYCLE: usize = 4;
+
+#[derive(Derivative, CSSelectable)]
+#[derivative(Clone, Debug)]
+pub struct EcAddPrecompileCallParams<F: SmallField> {
+    pub input_page: UInt32<F>,
+    pub input_offset: UInt32<F>,
+    pub output_page: UInt32<F>,
+    pub output_offset: UInt32<F>,
+}
+
+impl<F: SmallField> EcAddPrecompileCallParams<F> {
+    pub fn from_encoding<CS: ConstraintSystem<F>>(_cs: &mut CS, encoding: UInt256<F>) -> Self {
+        let input_offset = encoding.inner[0];
+        let output_offset = encoding.inner[2];
+        let input_page = encoding.inner[4];
+        let output_page = encoding.inner[5];
+
+        let new = Self {
+            input_page,
+            input_offset,
+            output_page,
+            output_offset,
+        };
+
+        new
+    }
+}
+
+fn ecadd_precompile_inner<F: SmallField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    x1: &UInt256<F>,
+    y1: &UInt256<F>,
+    x2: &UInt256<F>,
+    y2: &UInt256<F>,
+) -> (Boolean<F>, UInt256<F>, UInt256<F>) {
+    let base_field_params = &Arc::new(bn254_base_field_params());
+
+    // TODO: add validation of input values
+    let x1 = convert_uint256_to_field_element(cs, &x1, base_field_params);
+    let y1 = convert_uint256_to_field_element(cs, &y1, base_field_params);
+    let mut point1 = BN256SWProjectivePoint::from_xy_unchecked(cs, x1, y1);
+
+    let x2 = convert_uint256_to_field_element(cs, &x2, base_field_params);
+    let y2 = convert_uint256_to_field_element(cs, &y2, base_field_params);
```

Comment:
Do you mean validation of inputs? I can't quite get how it will work 👨‍💻

Reply:
oh, my bad - yeah it would be a bit more involved. i'm assuming the TODO means you want to ensure that x and y make a valid curve point

Reply:
yeah, it's just that in yul contracts we do have ready validation that the point is on curve, so this will be duplicating with worse performance; I would though wait for the review from era-contracts and see what guys would say - shall we leverage the checks here or no. Plus further integration I assume will tell a bit more on this subject

Reply:
makes sense - in this case doing pre validation in the contracts would be cheaper for sure